### PR TITLE
nat_ref for link road in French métropole

### PR DESCRIPTION
**Pull request by didier2020**

### CODE REVIEW

File: analysers/analyser_merge_ref_link_FR.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+## You should have received a copy of the GNU General Public License     ##
+## along with this program.  If not, see <http://www.gnu.org/licenses/>. ##
+##                                                                       ##
+###########################################################################
+
+from modules.OsmoseTranslation import T_
+from .Analyser_Merge import SHP, SourceDataGouv, Load, Select, Mapping
+from .Analyser_Merge_Network import Analyser_Merge_Network, ConflateNetwork
+
+class Analyser_Merge_Ref_Link_FR(Analyser_Merge_Network):
```
Comment (frodrigo) [inline analysers/analyser_merge_ref_link_FR.py:27]: Please rename class and file `analyser_merge_highway_link_ref_FR`

File: analysers/analyser_merge_highway_link_ref_FR.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+from .Analyser_Merge_Network import Analyser_Merge_Network, ConflateNetwork
+
+class Analyser_Merge_Highway_Link_Ref_FR(Analyser_Merge_Network):
+    def __init__(self, config, logger = None):
+        Analyser_Merge_Network.__init__(self, config, logger)
+        doc = dict(
+            detail = T_(
+'''The list of road comes from the database "RIU" in France.'''),
+            trap = T_(
+'''Those nat_ref can only be on `*_link` ways or roundabout.'''))
```
Comment (Famlam) [inline analysers/analyser_merge_highway_link_ref_FR.py:34]: roundabouts (plural)

**Comment by didier2020** [comment]
hello, 
there is an error in "real test"
'''
2024-08-14 09:25:50   error: error on analyse merge_highway_link_ref_FR...
2024-08-14 09:25:50     Traceback (most recent call last):
2024-08-14 09:25:50       File "/opt/osmose-backend/./osmose_run.py", line 244, in execc
2024-08-14 09:25:50         with obj(analyser_conf, logger.sub()) as analyser_obj:
2024-08-14 09:25:50              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-14 09:25:50       File "/opt/osmose-backend/analysers/analyser_merge_highway_link_ref_FR.py", line 36, in __init__
2024-08-14 09:25:50         self.def_class_possible_merge(item = 8432, id = 47, level = 3, tags = ['merge', 'highway'],
2024-08-14 09:25:50         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-14 09:25:50     AttributeError: 'Analyser_Merge_Highway_Link_Ref_FR' object has no attribute 'def_class_possible_merge'
'''

File: analysers/analyser_merge_highway_link_ref_FR.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+class Analyser_Merge_Highway_Link_Ref_FR(Analyser_Merge_Network):
+    def __init__(self, config, logger = None):
+        Analyser_Merge_Network.__init__(self, config, logger)
+        doc = dict(
+            detail = T_('''The list of road comes from the database "RIU" in France.'''),
+            trap = T_('''Those nat_ref can only be on `*_link` ways or roundabouts.'''))
+
+        self.def_class_possible_merge(item = 8432, id = 47, level = 3, tags = ['merge', 'highway'],
+            title = T_('*_link French métropole nat_ref integration suggestion'), **doc)
```
Comment (frodrigo) [inline analysers/analyser_merge_highway_link_ref_FR.py:35]: Use a more generic text. Like "*_link nat_ref integration suggestion", avoid change or duplicate translation.

File: analysers/analyser_merge_highway_link_ref_FR.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                    static2 = {"source:nat_ref": self.source},
+                    mapping1 = {
+                        "operator": 'gestionnai',
+                        "nat_ref": lambda row: (self.linkplo_isidor(row))}
+                )))
+
+        def is_riu_link(self,row):
+            # Filter only DB, FB
+            return row['nom_plo_in'][0:2] == 'DB'
+        def linkplo_isidor(self,row):
```
Comment (frodrigo) [inline analysers/analyser_merge_highway_link_ref_FR.py:66]:
```suggestion

        def linkplo_isidor(self,row):
```

File: analysers/analyser_merge_highway_link_ref_FR.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                dataset="57a837e2c751df5b90bb5dd4",
+                resource='90a65602-3ca4-41d7-bf7c-23d435c916e1'),
+                zip='VSMAP_TOUT.shp'),
+            Load('geom',
+                table_name = 'ref_link_fr_' + config.options['country'].replace("-", "_"),
+                where = lambda row: (self.is_riu_link(row))),
+            ConflateNetwork(
+                select = Select(
+                    types = ["ways"],
+                    tags = [{"highway": ["motorway_link","trunk_link","primary_link"],"junction":"roundabout"}]),
```
Comment (frodrigo) [inline analysers/analyser_merge_highway_link_ref_FR.py:52]: Multiline, please.

File: analysers/analyser_merge_highway_link_ref_FR.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                    mapping1 = {
+                        "operator": 'gestionnai',
+                        "nat_ref": lambda row: (self.linkplo_isidor(row))}
+                )))
+
+        def is_riu_link(self,row):
+            # Filter only DB, FB
+            return row['nom_plo_in'][0:2] == 'DB'
+        def linkplo_isidor(self,row):
+            #plo format isidor
```
Comment (frodrigo) [inline analysers/analyser_merge_highway_link_ref_FR.py:67]:
```suggestion
            # plo format isidor
```

Comment (didier2020) [reply analysers/analyser_merge_highway_link_ref_FR.py:67]: thanks

File: analysers/analyser_merge_highway_link_ref_FR.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        self.init(
+            "https://www.data.gouv.fr/fr/datasets/liaisons-du-reseau-routier-national/",
+            "Liaisons du réseau routier national",
+            SHP(SourceDataGouv(
+                attribution="data.gouv.fr:Ministère de la Transition écologique",
+                dataset="57a837e2c751df5b90bb5dd4",
+                resource='90a65602-3ca4-41d7-bf7c-23d435c916e1'),
+                zip='VSMAP_TOUT.shp'),
+            Load('geom',
+                table_name = 'ref_link_fr_' + config.options['country'].replace("-", "_"),
```
Comment (frodrigo) [inline analysers/analyser_merge_highway_link_ref_FR.py:47]: Why do you use a custom `table_name` with country name?

File: analysers/analyser_merge_highway_link_ref_FR.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                where = lambda row: (self.is_riu_link(row))),
```
Comment (frodrigo) [inline analysers/analyser_merge_highway_link_ref_FR.py:48]: As the condition is simple, you can use `select` in place of `where`. Select is done in the DB is faster.

File: analysers/analyser_merge_highway_link_ref_FR.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            ConflateNetwork(
+                select = Select(
+                    types = ["ways"],
+                    tags = [
+                        {"highway": ["motorway_link","trunk_link","primary_link"],
+                         "junction":"roundabout"}]),
+                osmRef = "nat_ref",
+                conflationDistance = 15,
+                minLength = 50,
```
Comment (frodrigo) [inline analysers/analyser_merge_highway_link_ref_FR.py:57]: What is `minLength` for? It is not a documented parameter.

File: analysers/analyser_merge_highway_link_ref_FR.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+                mapping = Mapping(
+                    static2 = {"source:nat_ref": self.source},
+                    mapping1 = {
+                        "operator": 'gestionnai',
+                        "nat_ref": lambda row: (self.linkplo_isidor(row))}
```
Comment (frodrigo) [inline analysers/analyser_merge_highway_link_ref_FR.py:62]:
```suggestion
                        "nat_ref": lambda row: self.linkplo_isidor(row)}
```

File: analysers/analyser_merge_highway_link_ref_FR.py (hunk 9 above)
Comment (didier2020) [reply analysers/analyser_merge_highway_link_ref_FR.py:57]: it's in Analyser_Merge_Network  line 95

File: analysers/analyser_merge_highway_link_ref_FR.py (hunk 7 above)
Comment (didier2020) [reply analysers/analyser_merge_highway_link_ref_FR.py:47]: to not use same like Analyser_Merge_Road_FR

File: analysers/analyser_merge_highway_link_ref_FR.py (hunk 9 above)
Comment (frodrigo) [reply analysers/analyser_merge_highway_link_ref_FR.py:58]: Ok, yes, it is a merge network.

File: analysers/analyser_merge_highway_link_ref_FR.py (hunk 7 above)
Comment (frodrigo) [reply analysers/analyser_merge_highway_link_ref_FR.py:47]: Non, in Analyser_Merge_Road_FR, all departement shp fila contains the same file and need to be rename. It is not your case.

**Comment by frodrigo** [comment]
Why did you need to add Analyser_Merge_Network_Link ?

**Comment by didier2020** [comment]
because of error def_class_possible_merge is not in Analyser_Merge_Network.

**Pull request closed without merging**